extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/caffeine/CaffeineCache.java

```diff
@@ -68,9 +68,15 @@ public CaffeineCache(CaffeineCacheInfo cacheInfo) {
         CompletableFuture<Object> newCacheValue = new CompletableFuture<Object>();
         CompletableFuture<Object> existingCacheValue = cache.asMap().putIfAbsent(key, newCacheValue);
         if (existingCacheValue == null) {
-            Object value = valueLoader.apply(key);
-            newCacheValue.complete(NullValueConverter.toCacheValue(value));
-            return unwrapCacheValue(newCacheValue);
+            try {
+                Object value = valueLoader.apply(key);
+                newCacheValue.complete(NullValueConverter.toCacheValue(value));
+                return unwrapCacheValue(newCacheValue);
+            } catch (Throwable t) {
+                cache.asMap().remove(key, newCacheValue);
+                newCacheValue.completeExceptionally(t);
```

Comment:
@ben-manes I need your help with this line. When it is executed, Caffeine is logging a warning about an `Exception thrown during asynchronous load` (the old warning I already got rid of before). Why is this happening while the future was removed from the cache with the previous line?

Reply:
I suppose because there is no way to remove a whenComplete callback. I didn’t think the put added the handler but would need to review the asMap code

Reply:
Yes, this is due to the callback and that cannot be removed by the CompletableFuture api. There is no mechanism for the callback to be aware of the situation for how a removal occurred and to no-op. At best it can safeguard against multiple invocations. The [code](https://github.com/ben-manes/caffeine/blob/b95222ab4fbf4b01f34ea053522227cc4e3ee3e2/caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalAsyncCache.java#L187-L214) in question.

The logic in this callback is to try to update the entry's metadata (weight, expiration) on success, try to remove it on failure, and record the statistics. If the entry was replaced or removed prior to completion then the conditional writes fail, but the statistics are still valid so they are recorded. For example an in-progress load that is removed still counts towards the load time and having started due to a cache miss. Similarly, `LoadingCache.refresh(key)` might load but fail to replace the entry if it was explicit removed, such as after when the system of record was updated and the cache invalidated.

Trying to special case the logger to sometimes log and other times not based on guessing the likely user interaction is error prone. If the use-case is to warn because users forget to handle errors, an ignoring to log when an in-flight future was explicitly inserted doesn't somehow negate error handling. I would agree the cache isn't responsible for a user's error handling, but unfortunately real world code means we have to be mindful that most will forget and silently ignore production errors.

I suspect the cleanest solution is simply to disable the logger. That is a simple configuration or api call and matches how loggers are intended to be used. Otherwise, you will have to restore the workaround to suppress the error from the cache's perspective in order to silence this message.

Reply:
Thanks for your very detailed answer @ben-manes!

I don't really like the idea of this extension changing the behavior of Caffeine application-wide, so I restored the workaround to suppress the warning.

Reply:
That is true, except I thought (maybe incorrectly) that Quarkus shades Caffeine so it would only apply to your extension if you disable the logger at the shaded path? I might be completely wrong, in which case you're very much right.

Reply:
You're probably right, I don't know everything about Quarkus and this specific feature is unknown to me. I'll see if I can find some information about it.

Reply:
We don't shade Caffeine.

Reply:
Thanks @gsmet! Then the current fix is the good one.